ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixFactory.java

```diff
@@ -69,10 +73,78 @@ public HelixManager getZkHelixManagerAndConnect(String clusterName, String insta
+   *               pool with this address, it will be reused.
+   * @param metrics the {@link DataNodeConfigSourceMetrics} to use.
+   * @return either a new instance of {@link DataNodeConfigSource} with the supplied configuration, or one from the pool
+   *         if there is already one created for this address.
+   */
+  public DataNodeConfigSource getDataNodeConfigSource(ClusterMapConfig clusterMapConfig, String zkAddr,
+      DataNodeConfigSourceMetrics metrics) {
+    return dataNodeConfigSources.computeIfAbsent(zkAddr,
+        k -> buildDataNodeConfigSource(clusterMapConfig, zkAddr, metrics));
+  }
+
+  /**
+   * Construct a new instance of {@link HelixManager}. Exposed so that tests can override if needed.
+   * @param clusterName the name of the cluster for the manager.
+   * @param instanceName the name of the instance for the manager.
+   * @param instanceType the {@link InstanceType} of the requester.
+   * @param zkAddr the address identifying the zk service to which this request is to be made.
+   * @return a new instance of {@link HelixManager}.
+   */
+  HelixManager buildZKHelixManager(String clusterName, String instanceName, InstanceType instanceType, String zkAddr) {
+    return HelixManagerFactory.getZKHelixManager(clusterName, instanceName, instanceType, zkAddr);
+  }
+
+  /**
+   * @param clusterMapConfig the {@link ClusterMapConfig} to use.
+   * @param zkAddr the ZooKeeper address to connect to. If a {@link HelixManager} is required and one is already in the
+   *               pool with this address, it will be reused.
+   * @param metrics the {@link DataNodeConfigSourceMetrics} to use.
+   * @return a new instance of {@link DataNodeConfigSource} with the supplied configuration.
+   */
+  private DataNodeConfigSource buildDataNodeConfigSource(ClusterMapConfig clusterMapConfig, String zkAddr,
+      DataNodeConfigSourceMetrics metrics) {
+    try {
+      InstanceConfigToDataNodeConfigAdapter instanceConfigSource = null;
+      if (clusterMapConfig.clusterMapDataNodeConfigSourceType.isInstanceConfigAware()) {
+        instanceConfigSource = new InstanceConfigToDataNodeConfigAdapter(
+            getZkHelixManagerAndConnect(clusterMapConfig.clusterMapClusterName,
+                ClusterMapUtils.getInstanceName(clusterMapConfig.clusterMapHostName, clusterMapConfig.clusterMapPort),
+                InstanceType.SPECTATOR, zkAddr), clusterMapConfig);
+      }
+      PropertyStoreToDataNodeConfigAdapter propertyStoreSource = null;
+      if (clusterMapConfig.clusterMapDataNodeConfigSourceType.isPropertyStoreAware()) {
+        propertyStoreSource = new PropertyStoreToDataNodeConfigAdapter(zkAddr, clusterMapConfig);
+      }
+
+      switch (clusterMapConfig.clusterMapDataNodeConfigSourceType) {
+        case INSTANCE_CONFIG:
+          return instanceConfigSource;
+        case PROPERTY_STORE:
+          return propertyStoreSource;
+        case COMPOSITE_INSTANCE_CONFIG_PRIMARY:
```

Comment:
Sure, I should publish a short doc on the migration steps actually.

These two both can be used as migration tools so that both property store and instance configs get updated.
The first step would be to switch to `COMPOSITE_INSTANCE_CONFIG_PRIMARY`, which would keep logic in ambry-server/ambry-frontend based on what the instance configs (legacy source) present, but also check for discrepancies and keep the property store up to date on updates to things like sealed partitions.

After we are comfortable that we aren't seeing discrepancies, we can switch to `COMPOSITE_PROPERTY_STORE_PRIMARY`. This will use the new source for application logic/listeners but still keep the instanceconfigs up to date in case we need to do a rollback.

Finally, one can switch to `PROPERTY_STORE` to complete the migration.